### PR TITLE
Berlin jun2019 presentation and misc fixes

### PR DESCRIPTION
**Pull request by hjorthjort**
The presentation and demo files from Berlin 2019 "Wasm on the blockchain" conference, plus some misc changes to semantics that was useful for the demo.

### CODE REVIEW

File: wasm.md [hunk 1]
```diff
@@ -254,7 +254,8 @@ Note that we do not need to call `#chop` on the results here.
 ```k
     syntax IBinOp ::= "div_u" | "rem_u"
  // -----------------------------------
-    rule <k> ITYPE . div_u I1 I2 => #if I2 =/=Int 0 #then < ITYPE > (I1 /Int I2) #else undefined #fi ... </k>
+    rule <k> ITYPE . div_u I1 I2 => < ITYPE > (I1 /Int I2)  ... </k> requires I2 =/=Int 0
+    rule <k> ITYPE . div_u I1 I2 => undefined ... </k>               requires I2  ==Int 0
```
Comment (MrChico) [inline wasm.md]: we should make a consistent choice on whether to branch on these instructions, in which case  `rem_{u,s}` and `div_s` should be updated as well

Comment (ehildenb) [reply wasm.md]: Branching (as this does here) is probably better.

File: wasm.md [hunk 2]
```diff
@@ -481,9 +482,10 @@ Note that, unlike in the WebAssembly specification document, we do not need the
     rule <k> ( br N ) ~> (IS:Instrs => .) ... </k>
     rule <k> ( br N ) ~> L:Label => #if N ==Int 0 #then L #else ( br N -Int 1 ) #fi ... </k>
 
-    syntax Instr ::= "(" "br_if" Int ")"
+    syntax Instr ::= "(" "br_if" Int ")" | "(" "br_if" Int Instr ")"
  // ------------------------------------
```
Comment (ehildenb) [inline wasm.md]: Formatting.

File: wasm.md [hunk 3]
```diff
@@ -542,17 +546,19 @@ The `*_local` instructions are defined here.
 
 ```k
     syntax Instr ::= "(" "local.get" Int ")"
-                   | "(" "local.set" Int ")"
-                   | "(" "local.tee" Int ")"
+                   | "(" "local.set" Int ")" | "(" "local.set" Int Instr ")"
+                   | "(" "local.tee" Int ")" | "(" "local.tee" Int Instr ")"
  // ----------------------------------------
```
Comment (ehildenb) [inline wasm.md]: Formatting.

File: wasm.md (hunk 1 above)
Comment (hjorthjort) [reply wasm.md]: At least it give a lot nicer behavior tree in KLab, which is why I changed it for `div_u` (it's the only one I used in the presentation). I'll change it for the others as well.

**Comment by hjorthjort** [comment]
@MrChico Would it make sense to merge `oldSchool` to `master` in KLab? I'm currently seeing a crash saying "can't read property `label` of `undefined`" or similar when using the master branch of KLab, guessing this was the fix:

```
-        let k = state.nodes[node.id.split("_")[0]];
-        if(k) {
-          node.oog = kast.get(k, "ethereum.evm.statusCode").label == "EVMC_OUT_OF_GAS_NETWORK"
-          node.gas = kast.get(k, "ethereum.evm.callState.gas")
-        }
+        // let k = state.nodes[node.id.split("_")[0]];
+        // if(k) {
+        //   node.oog = kast.get(k, "ethereum.evm.statusCode").label == "EVMC_OUT_OF_GAS_NETWORK"
+        //   node.gas = kast.get(k, "ethereum.evm.callState.gas")
+        // }
```

Maybe the original code should just be wrapped in some error handling? At least until KLab has a more modular structure with less EVM stuff hard coded.

File: wasm.md (hunk 2 above)
Comment (hjorthjort) [reply wasm.md]: Fixed.

File: wasm.md [hunk 4]
```diff
@@ -487,14 +492,16 @@ Note that, unlike in the WebAssembly specification document, we do not need the
     rule <k> ( br N ) ~> L:Label => ( br N -Int 1 ) ... </k>
       requires N >Int 0
 
-    syntax Instr ::= "(" "br_if" Int ")"
+    syntax Instr ::= "(" "br_if" Int ")" | "(" "br_if" Int Instr ")"
+ // ----------------------------------------------------------------
+    rule <k> ( br_if N I:Instr ) => I ~> ( br_if N ) ... </k>
  // ------------------------------------
```
Comment (ehildenb) [inline wasm.md]: formatting here is pretty weird

File: wasm.md [hunk 5]
```diff
@@ -256,26 +256,31 @@ Note that we do not need to call `#chop` on the results here.
 ```k
     syntax IBinOp ::= "div_u" | "rem_u"
  // -----------------------------------
-    rule <k> ITYPE . div_u I1 I2 => #if I2 =/=Int 0 #then < ITYPE > (I1 /Int I2) #else undefined #fi ... </k>
-    rule <k> ITYPE . rem_u I1 I2 => #if I2 =/=Int 0 #then < ITYPE > (I1 %Int I2) #else undefined #fi ... </k>
+    rule <k> ITYPE . div_u I1 I2 => < ITYPE > I1 /Int I2 ... </k> requires I2 =/=Int 0
+    rule <k> ITYPE . div_u I1 I2 => undefined ... </k>            requires I2  ==Int 0
```
Comment (ehildenb) [inline wasm.md]: Can we line up the `...`.

File: wasm.md [hunk 6]
```diff
@@ -256,26 +256,31 @@ Note that we do not need to call `#chop` on the results here.
...
+
+    rule <k> ITYPE . rem_u I1 I2 => < ITYPE > I1 %Int I2 ... </k> requires I2 =/=Int 0
+    rule <k> ITYPE . rem_u I1 I2 => undefined ... </k>            requires I2  ==Int 0
```
Comment (ehildenb) [inline wasm.md]: Also here, line up the `...`.

File: wasm.md (hunk 4 above)
Comment (hjorthjort) [reply wasm.md]: Fixed

File: wasm.md (hunk 5 above)
Comment (hjorthjort) [reply wasm.md]: Done

File: wasm.md (hunk 6 above)
Comment (hjorthjort) [reply wasm.md]: Done.

File: wasm.md [hunk 7]
```diff
@@ -256,26 +256,31 @@ Note that we do not need to call `#chop` on the results here.
...
-          => #if I2 =/=Int 0
-                #then < ITYPE > #unsigned(ITYPE, #signed(ITYPE, I1) %Int #signed(ITYPE, I2))
-                #else undefined
-             #fi
-         ...
-         </k>
+          => < ITYPE > #unsigned(ITYPE, #signed(ITYPE, I1) %Int #signed(ITYPE, I2)) ... </k>
+      requires I2 =/=Int 0
+       andBool I1 /Int I2 =/=Int #pow1(ITYPE)
+    rule <k> ITYPE . rem_s I1 I2 => undefined ... </k> requires I2 ==Int 0
```
[the remark this answers is not in the document]
Comment (hjorthjort) [reply wasm.md]: No, the `andBool` condition shouldn't be there. Good catch!

[the remark this answers is not in the document]
Comment (hjorthjort) [reply wasm.md]: Fixed. Also added regression tests.

**Pull request merged**